numpy/core/_internal.py

```diff
@@ -415,10 +327,17 @@ def _getfield_is_safe(oldtype, newtype, offset):
         If the field access is invalid
 
     """
-    new_fields = _get_all_field_offsets(newtype, offset)
-    old_fields = _get_all_field_offsets(oldtype)
-    # raises if there is a problem
-    _check_field_overlap(new_fields, old_fields)
+    if newtype.hasobject or oldtype.hasobject:
+        if offset == 0 and newtype == oldtype:
+            return
+
+        if oldtype.names:
+            for name in oldtype.names:
+                if (oldtype.fields[name][1] == offset and
```

Comment:
The idea is to allow getfield as long as we are getting a field that already exists. Eg, 

    >>> ones(3, dtype='i4,O').setfield(5, 'O', 4)

Since getfield doesn't know the name of the field we have to search for the field based on the offset.